src/upgrade.c

```diff
@@ -164,13 +170,14 @@ static int upgrade_main(void *ctx, struct apk_database *db, struct apk_string_ar
 	if (uctx->self_upgrade_only)
 		return 0;
 
-	if (uctx->ignore) {
-		char **pkg_name;
-		struct apk_name *name;
-		foreach_array_item(pkg_name, args) {
-			name = apk_db_get_name(db, APK_BLOB_STR(*pkg_name));
-			apk_solver_set_name_flags(name, solver_flags | APK_SOLVERF_IGNORE_UPGRADE, 0);
+	if (args->num) {
+		if (!(solver_flags & APK_SOLVERF_IGNORE_UPGRADE)) {
+			solver_flags = 0;
 		}
+
+		apk_name_foreach_matching(db, args, apk_foreach_genid(), set_solver_flags, ctx);
+
+		solver_flags &= ~APK_SOLVERF_IGNORE_UPGRADE;
```

Comment:
Since I removed the `ignore` bit from the upgrade context and directly updated the `solver_flags` in the option parsing as asked, I now have to reset this bit in the `solver_flags` after this part otherwise all the packages will be flagged with the ignore bit.